py/builtinimport.c: Remove partially-loaded modules from sys.modules.

DavidEGrayson opened the pull request:
This fixes #967 and my duplicate issue #11389.  It makes the behavior closer to CPython and makes it easier to debug Python code that the interpreter is failing to import.

stinos commented:
+1

Could you elaborate the commit message  instead of referring to the github issue only, so that when reading this it's clear what this is about? And since the commit literally says 'remove from sys.modules', it would be good to add that to the test as well.

dlech commented on py/builtinimport.c:
```diff
@@ -438,7 +439,14 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
 
-        if (stat == MP_IMPORT_STAT_DIR) {
+        nlr_buf_t nlr;
+        if (nlr_push(&nlr)) {
```
This doesn't look right. `nlr_push` is the equivalent of `try` in Python so would need to be before the code that can raise the exception.

DavidEGrayson replied:
This sounds like a code style issue.  I'll do whatever is needed to get this merged, but I chose this arrangement so that I wouldn't have to increase the indentation of the ~27 lines of code we are catching exceptions from, making it harder to fit code and comments there.

dlech replied:
This is not a style issue, but rather a functional one. The code as it is currently will remove all modules from `sys.modules`, not just ones that fail to import.

Furthermore, the error handling should be in the `else` statement of the `if (nlr_push...` and the `nlr_pop` should be at the end of the `if` block (right before `} else {`).

dlech commented on py/objmodule.c:
```diff
@@ -166,6 +166,11 @@ mp_obj_t mp_obj_new_module(qstr module_name) {
     return MP_OBJ_FROM_PTR(o);
 }
 
+void mp_obj_module_unregister(qstr module_name) {
```
Instead of adding a new function to unregister a module, would it make more sense to just not register it when it is created, but rather defer registering it until we are sure that it was successfully created?

Also, there is a comment in `mp_obj_new_module()` that says:

>      // We could error out if module already exists, but let C extensions
>      // add new members to existing modules.

Which means with the proposed unregister on error handling, we could actually end up unregistering a C module when trying to extend it if there was an error in the extension code. Not sure if that is a problem since the module wouldn't be usable anyway, but something to consider.

stinos replied:
> Instead of adding a new function to unregister a module, would it make more sense to just not register it when it is created, but rather defer registering it until we are sure that it was successfully created?

This could be tricky; I'm not sure but I seem to recall that for e.g. nested imports and errors therein to work correctly, the module should be present early already. On the other hand there should be enough tests which would reveal that to be the case. Code size might be smaller by not creating an extra function but just putting those 2 lines where needed.

DavidEGrayson replied:
I added a comment to the commit message and new test to justify why we still register the module early (circular imports).  

I'm not sure what specific scenario @dlech is talking about where my change would unregister a C module.  I guess there's some way to have a built-in module written in C and then a Python module with the same name that adds things to it?  I haven't come across that yet.

dlech replied:
> This could be tricky

Some other ideas from the related issue: https://github.com/micropython/micropython/issues/967#issuecomment-448461369

dlech commented on tests/import/import_broken.py:
Shouldn't the test print or otherwise verify `sys.modules` to ensure compatibility with CPython?

dlech commented on tests/import/import_broken.py:
```diff
@@ -0,0 +1,18 @@
+import sys
```
We usually don't use `assert` in MicroPython tests. Instead, we `print` and the output is compared to the same from CPython. So I would suggest a test like this:

```python
import sys

print("sys" in sys.modules)

try:
    from broken import x
except Exception as e:
    print(e.__class__.__name__)

# The broken module we tried to import should not be in sys.modules.
print("broken" in sys.modules)

# If we try to import the module again, the code should
# run again and we should get the same error.
try:
    from broken import x
except Exception as e:
    print(e.__class__.__name__)
```

dlech replied:
The `print("sys" in sys.modules)` line reveals a bug in the current implementation as I mentioned in another comment. (It's a little bit surprising to me that we don't have a test that catches this already.)

DavidEGrayson replied:
I misread your earlier message so that's why my response didn't really make sense.  You said `nlr_push` needs to be before the code that can raise the exception, and that is true.  The exception is raised within `do_load`, and every version of this pull request has called `nlr_push` before `do_load`.

The fact that `"sys"` is not in `sys.modules` is some other issue, which I'm not attempting to fix right now.  I added a test in `import_broken.py` that shows that normal, working modules are indeed added to `sys.modules`.

I went ahead and changed the code style the way you suggested though, putting the exception handling code *after* the code that throws the exception.  This requires about 27 lines of code to be indented, making it harder to fit code and comments in there, but I guess it is more readable.

DavidEGrayson commented:
I found that the old quirky behavior is documented in [tests/cpydiff/core_import_prereg.py](https://github.com/micropython/micropython/blob/786013d467d8beace7c7a8a2cd1122c799d05dae/tests/cpydiff/core_import_prereg.py), so I'll remove that file.

DavidEGrayson commented:
I think I've addressed all the comments from @dlech and @stinos.  Having some positive reviews of the code would be nice, and 
 probably help it get merged in sooner.

dlech commented on py/builtinimport.c:
```diff
@@ -438,31 +439,41 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
+            // An exception happened while running the module code above.
+            // Remove the partially-loaded module from sys.modules so future
+            // imports will attempt to run the code again.
+            mp_obj_module_unregister(full_mod_name);
```
Not creating/calling a new function like @stinos suggested saves about 50 bytes on x86_64 Linux.

```suggestion
            mp_obj_dict_delete(MP_OBJ_FROM_PTR(&MP_STATE_VM(mp_loaded_modules_dict)), MP_OBJ_NEW_QSTR(full_mod_name));
```

dpgeorge commented:
Thanks for working on this, it definitely needs to be fixed.  I'm actually surprised we lasted this long without such a fix (well, I guess in the normal case of imports succeeding you never hit this issue).

The main concern is the amount of C stack that is used by the NLR buffer.  It's 64 bytes on Cortex-M MCUs with a floating-point unit.  And C stack is a precious thing.  I have an idea how to improve this using NLR callbacks.

Also, it would be good to have a few more tests, eg for:
1. `a` imports `b`, then `a` fails`, but `b` should still remain imported
2. `a` imports `b`, but `b` fails, so neither modules should remain in `sys.modules`

dpgeorge commented:
> The main concern is the amount of C stack that is used by the NLR buffer. It's 64 bytes on Cortex-M MCUs with a floating-point unit. And C stack is a precious thing. I have an idea how to improve this using NLR callbacks.

See #11445 for that.

robert-hh commented:
>  And C stack is a precious thing.

@dpgeorge Since you mentioned it: what is a good C stack size for Micropython? 4k? 8k? or more? Threads use 4k by default.

dpgeorge commented:
> what is a good C stack size for Micropython?

It depends a lot on the application and how many nested Python calls it needs.  4k is OK, 8k I would say is standard, 16k for when you have memory to spare.

DavidEGrayson commented:
OK, I added the test cases suggested by @dpgeorge.  Yeah, using up 64 bytes of stack space for every nested import on a Cortex-M is unfortunate, so it sounds like we should figure out how to make some sort of list of callbacks that get called when a non-local return happens.

dpgeorge commented:
@DavidEGrayson now that #11445 has been merged, are you able to rebase/rework this PR to use the new NLR callback mechanism?

DavidEGrayson committed:
py/builtinimport.c: Remove partially-loaded modules from sys.modules.

This fixes issue #967: importing a module with a syntax error or some
other problem that happens at import time would result in a
potentially-incomplete module object getting added to sys.modules.
Subsequent imports would use that object, resulting in confusing
error messages that hide the root cause of the problem.

It is still important to add the module to sys.modules while the
import is happening so that we can support circular imports just
like CPython does.

Signed-off-by: David Grayson <davidegrayson@gmail.com>

DavidEGrayson commented:
Commit 265066f17 implements an NLR callback using GCC's [nested function](https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html) feature, which eliminates a struct and simplifies the code.  I'm expecting everyone to object to using GCC extensions, but I just want to see if we can get a code size report about it.

DavidEGrayson commented:
Code size [report](https://github.com/micropython/micropython/actions/runs/5172773128/jobs/9317431605) for this PR if we use GCC nested functions (commit 265066f179):

```text
   bare-arm:    +0 +0.000% 
minimal x86:  +168 +0.090% 
   unix x64:  +216 +0.027% standard
      stm32:  +108 +0.028% PYBV10
        rp2:  +144 +0.044% PICO
```

Code size [report](https://github.com/micropython/micropython/actions/runs/5172861481/jobs/9317605355?pr=11421) for this PR if we use standard C (commit 394d95e3f):

```text
   bare-arm:    +0 +0.000% 
minimal x86:  +112 +0.060% 
   unix x64:  +120 +0.015% standard
      stm32:   +56 +0.014% PYBV10
        rp2:   +56 +0.017% PICO
```

This is weird to me.  It looks like GCC does not comply with [part 2 of the C++ zero-overhead principle](https://en.cppreference.com/w/cpp/language/Zero-overhead_principle#:~:text=What%20you%20do%20use%20is%20just%20as%20efficient%20as%20what%20you%20could%20reasonably%20write%20by%20hand.).

It does make it a no-brainer to just use standard C.

dpgeorge commented on py/builtinimport.c:
```diff
@@ -31,6 +31,7 @@
 #include <assert.h>
 
 #include "py/compile.h"
+#include "py/nlr.h"
```
This include shouldn't be necessary (no other .c file in py/ includes it).

jimmo commented:
@DavidEGrayson That is very interesting, thanks for sharing! We cannot use nested functions because they're a (very nifty!) GNU extension but we need to support non-GCC compilers.

> which eliminates a struct

structs don't really 'exist' in the generated code so there's no code size benefit, but I'm guessing you meant more about simplification, which it certainly does!

The code size diff is interesting. I'd love to know if this is a missed optimisation (which can probably be forgive by the difficulity of scope / escape analysis), or something fundamental about the way the trampoline mechanism works (which is required because you're taking the address of the nested function).

And this is a good fix, something that has tripped up many people over the years, thanks for sending the PR!

dpgeorge commented on py/builtinimport.c:
```diff
@@ -346,6 +347,16 @@ STATIC void evaluate_relative_import(mp_int_t level, const char **module_name, s
+
+STATIC void unregister_module_from_nlr_jump_callback(void *ctx_in) {
+    nlr_jump_callback_node_unregister_module_t *ctx = ctx_in;
+    mp_obj_module_unregister(ctx->full_mod_name);
```
It would reduce code size (when LTO is not used) to inline `mp_obj_module_unregister()` directly here.

Also, no need to add more to the "public" API if it's not necessary (ie no need to add a new public function `mp_obj_module_unregister`).

If it's ever needed as a separate function, it can be factored out.

DavidEGrayson replied:
OK.  For some reason I thought `mp_obj_module_unregister` was using private data, but you're right, it can be inlined.

dpgeorge commented on py/builtinimport.c:
```diff
@@ -491,6 +503,12 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
+    // exception while running its code.
+    nlr_jump_callback_node_unregister_module_t ctx;
+    ctx.full_mod_name = full_mod_name;
+    nlr_push_jump_callback(&ctx.callback, unregister_module_from_nlr_jump_callback);
```
Should this be move earlier, in case the above `mp_obj_dict_store` fail?

DavidEGrayson replied:
Yeah, it probably makes sense for the module to be removed if anything goes wrong during the rest of the `process_import_at_level` function after the module was registered.

dpgeorge commented on py/builtinimport.c:
```diff
@@ -521,6 +539,8 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
         // a __path__ attribute, and not attempt to stat it.
     }
 
+    nlr_pop_jump_callback(false);
```
Should this be moved later, in case the `mp_store_attr` below fails (due to out-of-memory)?  Not sure...

DavidEGrayson replied:
Yeah, it probably makes sense for the module to be removed if anything goes wrong during the rest of the `process_import_at_level` function.

jimmo commented:
OK I was sufficiently interested to at least look at the generated assembly. (built for cortex m4 with `-Os`).

nested1.c  (doesn't use nested functions)
```
typedef struct {
    int a;
} callback_t;

int x(callback_t *t) {
    return t->a * t->a;
}

callback_t *tt;
int (*cc)(callback_t *);
void __attribute((noinline)) foo(callback_t *t, int (*c)(callback_t *)) {
    tt = t;
    cc = c;
}

int A(int a) {
    callback_t t;
    t.a = a;
    foo(&t, &x);
}
```

--->
```
00008000 <x>:
    8000:	6800      	ldr	r0, [r0, #0]
    8002:	4340      	muls	r0, r0
    8004:	4770      	bx	lr

00008006 <foo>:
    8006:	4b02      	ldr	r3, [pc, #8]	@ (8010 <foo+0xa>)
    8008:	e9c3 0100 	strd	r0, r1, [r3]
    800c:	4770      	bx	lr
    800e:	bf00      	nop
    8010:	00009028 	.word	0x00009028

00008014 <A>:
    8014:	b507      	push	{r0, r1, r2, lr}
    8016:	4903      	ldr	r1, [pc, #12]	@ (8024 <A+0x10>)
    8018:	a801      	add	r0, sp, #4
    801a:	f7ff fff4 	bl	8006 <foo>
    801e:	b003      	add	sp, #12
    8020:	f85d fb04 	ldr.w	pc, [sp], #4
    8024:	00008001 	.word	0x00008001
```

nested2.c (uses nested function)
```
typedef struct {
} callback_t;

callback_t *tt;
int (*cc)(callback_t *);
void __attribute((noinline)) foo(callback_t *t, int (*c)(callback_t *)) {
    tt = t;
    cc = c;
}

int A(int a) {
    int x(callback_t *) {
        return a * a;
    }
    callback_t t;
    foo(&t, &x);
}
```

--->
```
00008000 <x.0>:
    8000:	f8dc 0000 	ldr.w	r0, [ip]
    8004:	4340      	muls	r0, r0
    8006:	4770      	bx	lr

00008008 <foo>:
    8008:	4b01      	ldr	r3, [pc, #4]	@ (8010 <foo+0x8>)
    800a:	e9c3 0100 	strd	r0, r1, [r3]
    800e:	4770      	bx	lr
    8010:	00009058 	.word	0x00009058

00008014 <A>:
    8014:	b510      	push	{r4, lr}
    8016:	b088      	sub	sp, #32
    8018:	ab0a      	add	r3, sp, #40	@ 0x28
    801a:	9307      	str	r3, [sp, #28]
    801c:	4b08      	ldr	r3, [pc, #32]	@ (8040 <A+0x2c>)
    801e:	9002      	str	r0, [sp, #8]
    8020:	cb0f      	ldmia	r3, {r0, r1, r2, r3}
    8022:	ac03      	add	r4, sp, #12
    8024:	e884 000f 	stmia.w	r4, {r0, r1, r2, r3}
    8028:	ab02      	add	r3, sp, #8
    802a:	a801      	add	r0, sp, #4
    802c:	60a3      	str	r3, [r4, #8]
    802e:	f044 0101 	orr.w	r1, r4, #1
    8032:	4b04      	ldr	r3, [pc, #16]	@ (8044 <A+0x30>)
    8034:	60e3      	str	r3, [r4, #12]
    8036:	f7ff ffe7 	bl	8008 <foo>
    803a:	b008      	add	sp, #32
    803c:	bd10      	pop	{r4, pc}
    803e:	bf00      	nop
    8040:	00008048 	.word	0x00008048
    8044:	00008001 	.word	0x00008001
```

dpgeorge commented:
Thanks for updating, this looks really good now, and is only a small increase in code size.  And fixes a long-outstanding issue.

Rebased and merged in a79a6ab3641e3536e509278706e7ac4de01a8126 with a slightly tweaked commit message.
